### ISSUE TITLE
Implement writable streams for natively streaming inserts

### ISSUE THREAD

**Issue by metasloth** [googleapis/nodejs-bigquery#506]
**Is your feature request related to a problem? Please describe.**
It would be really useful to be able to natively stream records directly into a table (via `/insertAll`). This would make it much more accessible to transfer significant amounts of data on low-resource platforms like cloud functions.

 **Describe the solution you'd like**
A Table method similar to [createWriteStream()](https://googleapis.dev/nodejs/storage/latest/File.html#createWriteStream) in the @google-cloud/storage package. Being able to just call `mySourceStream.pipe(myTable.createWriteStream())` would be super helpful.

 **Describe alternatives you've considered**
I've managed to get roughly the functionality I'm looking for by calling `insert()` in arbitrarily sized batches, i.e.

```js
class TableWriteableStream extends Writable {
  constructor (destinationTable) {
    super({ objectMode: true })
    this.destinationTable = destinationTable
    this.records = []
  }

  _write (chunk, encoding, callback) {
    this.records.push(chunk)
    // Every 500 records, call insert()
    if (this.records.length >= 500) {
      const recordsToInsert = this.records
      this.records = []
      this.destinationTable
        .insert(recordsToInsert)
        .then(() => { callback() })
    } else {
      callback()
    }
  }

  _final (callback) {
      // Insert any remaining records ...

```

 **Additional context**
This is somewhat related to #75, though that seems focused on compression (which would be a great option for this feature, just like the 'gzip' option for the @google-cloud/storage function).

**Comment by callmehiphop** [comment]
I think this would make a pretty good addition to the client. We already have a `table.createWriteStream()` method that streams into a load job, so this would need to be called something else (like `table.createInsertStream()` maybe?). 

Since we would be piggy backing off of the `insert` API, I imagine this would be limited to writing only JSON.

```js
createRowStream()
  .pipe(table.createInsertStream({
    // should accept all valid insert options
    ignoreUnknownValues: true,
    // allow users to configure batch size
    maxRecords: 1000,
    // allow users to configure stream options
    streamOptions: {highWaterMark: 16},
  }))
  .on('error', err => {
    // err would be an API error
  })
  .on('response', resp => {
    // resp would be the full insertAll response from the most recent batch
  })
  .on('finish', () => {
    // fires when the stream is done
  });
```

@tswast @shollyman @bcoe any thoughts/feelings here?

**Comment by richardscarrott** [comment]
@callmehiphop is there any documentation on the difference between streaming data into BigQuery with `table.createWriteStream` vs `table.insert`; my understanding is the former creates a "load job" whereas the latter does not -- I can't find much information on what this means in practical terms?

Additionally, I think one reason I struggle to understand this is the fact that BigQuery streaming !== Node streaming so even though `table.insert` is not a streaming API in Node terms, it is a streaming API in BigQuery terms?

Aside, if you were to expose a writable insert API, I don't think it should be the responsibility of this client to batch as there's loads of existing packages which do this, e.g.

```js
import BatchStream from 'batch-stream';

createReadableStream()
  .pipe(new BatchStream({ size : 1000 }))
  .pipe(table.createInsertStream());
```

Or using highlandjs:
```js
import _ from 'highland';

_(createReadableStream())
  .batch(1000)
  .pipe(table.createInsertStream());
```

**Comment by DominicTobias** [comment]
Totally agree I've jumped through hoops trying to get `insertAll` to behave properly as it seems to cancel the last batch if it didn't finish in time. Tried to implement as a stream of queued promises but it's still erroring. Seems like a pretty fundamental thing to have for a BigQuery DB, not assuming we are using files to stream in. Didn't find https://cloud.google.com/solutions/streaming-cloud-pub-sub-messages-over-websockets helpful I don't want to implement a filesystem or pub/sub just stream from a high frequency websocket.
